[main] Update documentation for kubelet drop-in config dir

sohankunkerkar opened the pull request:
Xref: https://github.com/kubernetes/enhancements/pull/4419

We are planning to move the Kubelet drop-in configuration directory feature to Beta for v1.30. This PR makes the necessary documentation changes to accurately reflect the current state of the feature

sohankunkerkar commented:
cc @haircommander @SergeyKanzhelev

haircommander commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,11 +86,8 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
 
-As of Kubernetes v1.28.0, the kubelet has been extended to support a drop-in configuration directory. The location of it can be specified with
-`--config-dir` flag, and it defaults to `""`, or disabled, by default.
+In Kubernetes v1.30.0, the drop-in configuration directory feature has reached Beta status. This allows the kubelet to support a drop-in configuration directory, which can be specified using the `--config-dir` flag. By default, it is set to `""`, indicating that the feature is disabled by default.
```
I think this should be 1.29.0 and update with the new correct behavior. We can update to beta once we actually update the code to beta

sohankunkerkar replied:
Fixed

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,11 +86,8 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
 
-As of Kubernetes v1.28.0, the kubelet has been extended to support a drop-in configuration directory. The location of it can be specified with
-`--config-dir` flag, and it defaults to `""`, or disabled, by default.
+In Kubernetes v1.29.0, the drop-in configuration directory feature allows the kubelet to support a drop-in configuration directory, which can be specified using the `--config-dir` flag. By default, it is set to `""`, indicating that the feature is disabled by default. Users need to expliclty opt-in for this feature by setting the `--config-dir` flag to a valid directory path.
```
```suggestion
You can specify a drop-in configuration directory for the kubelet. By default, the kubelet does not look
for drop in configuration files anywhere; you must specify a path.
For example: `--config-dir=/etc/kubernetes/kubelet.conf.d`
```

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,11 +86,8 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
-As of Kubernetes v1.28.0, the kubelet has been extended to support a drop-in configuration directory. The location of it can be specified with
-`--config-dir` flag, and it defaults to `""`, or disabled, by default.
+In Kubernetes v1.29.0, the drop-in configuration directory feature allows the kubelet to support a drop-in configuration directory, which can be specified using the `--config-dir` flag. By default, it is set to `""`, indicating that the feature is disabled by default. Users need to expliclty opt-in for this feature by setting the `--config-dir` flag to a valid directory path.
 
```
```suggestion

For Kubernetes v1.28 to v{{< skew currentVersion >}}, you  can only specify `--config-dir` if you also set
the environment variable `KUBELET_CONFIG_DROPIN_DIR_ALPHA` for the kubelet process (the value
of that variable does not matter).
```

This change is appropriate for a PR targeting _main_; it will render right. Then, once _this_ is merged to main and has been onwards merged to _dev-1.30_, update that branch to mark the feature as beta. You would drop the
`v1.28 to v{{< skew currentVersion >}}` text and code it as (eg) `v1.28 and v1.29`.

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,11 +86,8 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
 
 ## Drop-in directory for kubelet configuration files {#kubelet-conf-d}
 
```
```suggestion

{{<feature-state for_k8s_version="v1.28" state="alpha" >}}

```

(To mark when the feature went alpha)

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,11 +86,8 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
-For Kubernetes v{{< skew currentVersion >}}, the kubelet returns an error if you specify `--config-dir` without that variable set, and startup fails.
 You cannot specify the drop-in configuration directory using the kubelet configuration file; only the CLI argument `--config-dir` can set it.
 
 One can use the kubelet configuration directory in a similar way to the kubelet config file.
```
(nit / cleanup)
```suggestion
You can use the kubelet configuration directory in a similar way to the kubelet configuration file.
```

sohankunkerkar commented:
> /sig node
> 
> I'm not sure about intent, but: tidying the live (v1.29) docs _is_ helpful.
> 
> I've suggested changes that I think are appropriate for a PR targeting _main_; it will render right. Then, once _this_ is merged to main and has been onwards merged to _dev-1.30_, send in a **second PR** to promote the feature to beta. The second PR should have _dev-1.30_ as its base branch.

@sftim Thanks for the suggestions. I have addressed your concerns in the latest push. PTAL

haircommander commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,14 +86,17 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+For example: `--config-dir=/etc/kubernetes/kubelet.conf.d`
 
-One can use the kubelet configuration directory in a similar way to the kubelet config file.
+For Kubernetes v1.28 to v{{< skew currentVersion >}}, you  can only specify `--config-dir` if you also set
```
wouldn't this be `v1.28 to v1.29` (instead of currentVersion, which would be 1.30)

sohankunkerkar replied:
I followed [this](https://github.com/kubernetes/website/pull/44985#discussion_r1475839135) comment

sftim replied:
The current version of Kubernetes is v1.29. 

> I've suggested changes that I think are appropriate for a PR targeting _main_; it will render right. Then, once this is merged to _main_ and has been onwards merged to _dev-1.30_, send in a second PR to promote the feature to beta. The second PR should have _dev-1.30_ as its base branch.

The approach I've suggest would work even if the k/k update to kubelet drop-in config didn't land until a later minor release.

sohankunkerkar replied:
@sftim does the latest push make sense?

sftim replied:
Yep

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,14 +86,17 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+the environment variable `KUBELET_CONFIG_DROPIN_DIR_ALPHA` for the kubelet process (the value
+of that variable does not matter).
+
+You can use the kubelet configuration directory in a similar way to the kubelet configuration file.
```
nit: I'd put a whilespace line after this one.

sohankunkerkar replied:
Fixed

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -117,12 +120,72 @@ kind: KubeletConfiguration
+* Command line arguments (highest precedence).
+
+{{< note >}}
+Kubeadm utilizes a specific [patching strategy](/docs/setup/production-environment/tools/kubeadm/control-plane-flags.md/#patches), where configuration files are merged based on an optional file name suffix, determining the order in an alphanumeric manner. Conversely, the kubelet config drop-in directory feature processes these configuration files in lexical order based on file names, i.e., config ...
```
How about:
```suggestion
The kubelet processes files in its config drop-in directory using a similar mechanism, sorting the
**entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration
files with a later sort order can override earlier files.

This offers you flexibility in how you manage and combine kubelet configuration that comes from
different sources.

---

The config drop in mechanism for the kubelet is similar but different from how the `kubeadm`
tool allows you to patch configuration. The `kubeadm` tool uses a specific
[patching strategy](/docs/setup/production-environment/tools/kubeadm/control-plane-flags/#patches)
for its configuration, where configuration files are merged based on an optional file name suffix.
The kubelet determines the order of merges based on sorting the **suffixes** alphanumerically.
```

sftim replied:
(this might be long enough to merit being an item with its own heading, rather than a note)

sohankunkerkar replied:
Let me know if the latest push makes sense here.

sohankunkerkar replied:
@sftim could you PTAL?

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -117,12 +121,74 @@ kind: KubeletConfiguration
+* The kubelet configuration.
 * Drop-in configuration files, according to sort order.
-* Feature gates specified over the command line (highest precedence).
+* Command line arguments (highest precedence).
```
it's hard to understand how feature gates from the command line has lowest precedence and command line - highest? Maybe say here that this excludes feature gates

sohankunkerkar replied:
Sure, I will add that information.

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -117,12 +121,74 @@ kind: KubeletConfiguration
+  - "192.168.0.3"
+  - "192.168.0.5"
+featureGates:
+  KubeletTracing: false
```
can the feature gate be overriden? Will there be an error or we will follow the priority order to decide on the final value?

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -117,12 +121,74 @@ kind: KubeletConfiguration
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: KubeletConfiguration
+clusterDNS:
+  - "192.168.0.2"
```
is there a way to clean up the list? Is it a typical requirement?

sohankunkerkar replied:
Not sure if I understood your question. Let me take a shot. When you refer to the main config, you are talking about the base kubelet config. The intention behind adding drop-in files is to modify or clean up the values in the base configuration.

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -117,12 +121,74 @@ kind: KubeletConfiguration
+kind: KubeletConfiguration
+clusterDNS:
+  - "192.168.0.2"
+  - "192.168.0.3"
```
can the previously defined single value be removed from the list?

sohankunkerkar replied:
Here's the caveat: you cannot remove a single value from the list. You need to replace the entire list with the new values. If you want to maintain the original values as well, you need to mention that in the new drop-in config.

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -117,12 +121,74 @@ kind: KubeletConfiguration
+  memory.available: "200Mi"
+address: "192.168.0.8"
+```
+Additionally, you can change the slices/lists values of the kubelet configuration. For example, you can override the `clusterDNS` and `featureGates` fields as follows:
````
please give an example with overriding values

SergeyKanzhelev replied:
example: `failSwapOn ` - disabled explicitly in main config and enabled in child config. May be a common scenario to enable swap via drop-in config

sohankunkerkar replied:
This is covered in the first example.

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -117,12 +121,74 @@ kind: KubeletConfiguration
+```yaml
+apiVersion: kubelet.config.k8s.io/v1beta1
+kind: KubeletConfiguration
+clusterDNS:
````
how eviction collections will be merged? Will the whole structure be overriden or one value added?

SergeyKanzhelev replied:
Same question about logging - will the whole structure be overriden or individual fields?

sohankunkerkar replied:
The merging logic is targeted for the individual fields. If the intention is to override fields, then mention only those fields in the drop-in configs.

SergeyKanzhelev replied:
individual high level fields or all of them? I simply don't know how it works and this documentation doesn't give a clear understanding.

SergeyKanzhelev replied:
evictionHard and evictionSoft is coming from - what if I want to configur MemorySwap. I need to enable feature gate, and potentially want to disable eviction hard and evictionSoft. Is it possible to do with a single drop-in file?

sohankunkerkar replied:
>evictionHard and evictionSoft is coming from - what if I want to configur MemorySwap. I need to enable feature gate, and potentially want to disable eviction hard and evictionSoft. Is it possible to do with a single drop-in file?

Yes, you can set the values in a drop-in config for this use case.

sohankunkerkar replied:
>individual high level fields or all of them? I simply don't know how it works and this documentation doesn't give a clear understanding.

Please scratch the previous response.

>how eviction collections will be merged? Will the whole structure be overriden or one value added?

In this case, the entire structure will be overridden. If you want to change just one field, you can certainly do that. For example, in the case of `evictionHard`, you can alter the `memory.available` to a new value. However, you have to set values for other fields, perhaps using default values or any other values you desire, to make that happen.

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,14 +86,18 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
 
 ## Drop-in directory for kubelet configuration files {#kubelet-conf-d}
```
Beta graduation criteria ask to document the /configz endpoint as this is a recommended and only way for a cluster admin to review the resuting configuration file. Can you please send a PR here for this?

Link to graduation criteria: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3983-drop-in-configuration/README.md#beta

sohankunkerkar replied:
Sure, does it make sense to add that before the drop-in directory point with the title `Viewing the kubelet configuration`?

SergeyKanzhelev replied:
this is the first mention of `/configz` in docs ever. I think it deserves it's own page. @sftim what would be the best place to document the `/configz` endopint?

sftim replied:
Hi. Best place would be within https://kubernetes.io/docs/reference/node/

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -117,12 +122,75 @@ kind: KubeletConfiguration
+  - "192.168.0.3"
+  - "192.168.0.5"
+featureGates:
+  MemoryQoS: false
```
if you didn't specify MemoryQoS in the drop in config - will it stay true or will change to whatever is default in k8s?

sohankunkerkar replied:
If you don't mention the `MemoryQoS` field in the drop-in config, it will display the value mentioned in the base kubelet config unless you set the value via a command-line flag, which takes higher precedence over the drop-in config values.

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -117,12 +122,75 @@ kind: KubeletConfiguration
+  - "192.168.0.2"
+  - "192.168.0.3"
+  - "192.168.0.5"
+featureGates:
```
will the speciall `all/alpha=TRUE` work in drop-in config? Will it respect previously disabled alphas or will force enable them?

sohankunkerkar replied:
>will the speciall all/alpha=TRUE work in drop-in config

It should work

>Will it respect previously disabled alphas or will force enable them?

The latter

mrunalp replied:
@sohankunkerkar can you include this in the website example?

sohankunkerkar replied:
>Will it respect previously disabled alphas or will force enable them?

The above logic will be handled by the Kubelet. This functionality aims to replace the corresponding field value.

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -117,12 +122,75 @@ kind: KubeletConfiguration
+serializeImagePulls: false
+evictionHard:
+  memory.available: "200Mi"
+clusterDNS:
```
My understanding is that all lists are being overriden as a whole. Not merged. If so - will it be the problem with this specific field? Will one ever want to simply add a new DNS via the drop-in config?

sohankunkerkar replied:
Patching is not currently supported in the existing implementation. Therefore, while adding a new DNS value is possible, it results in the removal of previous values in the `clusterDNS` list.

SergeyKanzhelev commented:
My understanding is that today we have a limitation:
- we do not allow to merge lists
- we do not allow to remove values from lists
- we do not allow to set a field value to "nothing" to force Kubelet's default

It will be great if those limitations will be clearly stated on the top of the documentation. And it will be clear if something is a list or not. For example, is the FeatureGates a list or set of fields? How the cluster admin may reason about it?

Can you please analyze fields in the node config on whether those limitations will block any scenarios that we may consider popular. Also, if there are any combinations of fields that depend on each other and may not work well? For example, one field enables something and another configures it. And if drop-in will disable it, the config part will break things unless explicitly unset.

Some examples I can think of:

- drop in config to enable verbose logging
- drop in config to enable swap: set failonSwap to false, set a single feature gate to true without overriding others, disable eviction on memory without touching any other configured evictions.
- changing the # of pods per node: increase the pods per node value AND increase the kubeReserved's memory value without touching other values
- Adding unsafe sysctls - various drop-in files to add unsafe sysctls. Will it only be possible to add all of them art once? Is it accepteable?

sohankunkerkar commented:
> My understanding is that today we have a limitation:
> 
> * we do not allow to merge lists
> * we do not allow to remove values from lists
> * we do not allow to set a field value to "nothing" to force Kubelet's default
> 
> It will be great if those limitations will be clearly stated on the top of the documentation.

Sure, let me go ahead and update the documentation to clarify the limitations.

> And it will be clear if something is a list or not. For example, is the FeatureGates a list or set of fields? How the cluster admin may reason about it?

I'm not sure if this is within the scope of the current implementation. We can defer to [this](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) documentation if the cluster admin encounters any issues.

sohankunkerkar commented:
> Can you please analyze fields in the node config on whether those limitations will block any scenarios that we may consider popular. Also, if there are any combinations of fields that depend on each other and may not work well? For example, one field enables something and another configures it. And if drop-in will disable it, the config part will break things unless explicitly unset.
> 
> Some examples I can think of:
> 
> * drop in config to enable verbose logging
> * drop in config to enable swap: set failonSwap to false, set a single feature gate to true without overriding others, disable eviction on memory without touching any other configured evictions.
> * changing the # of pods per node: increase the pods per node value AND increase the kubeReserved's memory value without touching other values
> * Adding unsafe sysctls - various drop-in files to add unsafe sysctls. Will it only be possible to add all of them art once? Is it accepteable?

:+1:

Thanks for the suggestions. I was wondering if this is considered a gating criterion to promote this feature to Beta.

sohankunkerkar commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -84,45 +84,309 @@ If `--config` is provided and the values are not specified via the command line,
+
+```bash
+$ curl -X GET http://127.0.0.1:8001/api/v1/nodes/<node-name>/proxy/configz | jq .
+{
````
Let me know if displaying the entire kubelet config makes sense here.

sohankunkerkar commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -84,45 +84,309 @@ If `--config` is provided and the values are not specified via the command line,
 
-As of Kubernetes v1.28.0, the kubelet has been extended to support a drop-in configuration directory. The location of it can be specified with
-`--config-dir` flag, and it defaults to `""`, or disabled, by default.
+{{<feature-state for_k8s_version="v1.28" state="alpha" >}}
```
https://github.com/kubernetes/website/pull/44985#discussion_r1480073608

sohankunkerkar replied:
This change is specific to Alpha and will be merged into the main branch. Subsequently, I will create another PR targeting the `1.30-dev` branch for transitioning to beta.

kannon92 replied:
Ah okay. Missed that.

kannon92 commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -84,45 +84,309 @@ If `--config` is provided and the values are not specified via the command line,
+
+To inspect the kubelet configuration, follow these steps:
+
+1. Start a proxy server using `kubectl proxy` in your terminal.
```
Is this code always going to work? 

I don't know much about `kubectl proxy` but this seems that it would be locked down from most users using managed cloud services?

And do we need to pass any certs or tls for curl for this to work?

sohankunkerkar replied:
There's another way to query the kubelet config
```bash
$ kubectl get --raw "/api/v1/nodes/<node-name>/proxy/configz" | jq
```
which doesn't involve passing any certs,

sohankunkerkar replied:
Yeah, that's tricky. I wrote this documentation with the normal use case in mind.

kannon92 replied:
I don't know what normal would be in this case. I would usually suggest mentioning links to `kubectl proxy` or something more detailed in case it doesn't work for the user..

mrunalp commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -84,45 +84,309 @@ If `--config` is provided and the values are not specified via the command line,
+  KubeletTracing: true
+staticPodURLHeader:
+  kubelet-api-support:
+  - "Authorization: 8945AFSG1"
```
Can you show an example of dropping or not specififying a field in the override file?

sohankunkerkar replied:
Fixed

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -84,45 +84,309 @@ If `--config` is provided and the values are not specified via the command line,
 defaults for the `KubeletConfiguration` version apply.
 In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
 
+## Viewing the kubelet configuration
```
what I am looking for when asking for /configz documentation is a clear statement on what that config represents:

1. Does it merge-in all the command line arguments into it?
2. Does it merge-in all the drop-in configuration files?
3.  Does it populate all the defaults?

In the example in this section, the feature gate consist of a single `"AllAlpha": false`. I wonder if it was configured by you explicitly or this is what it returns out of the box with nothing pre-configured?

sohankunkerkar replied:
>what I am looking for when asking for /configz documentation is a clear statement on what that config represents:

>    Does it merge-in all the command line arguments into it?
>   Does it merge-in all the drop-in configuration files?
>  Does it populate all the defaults?

Do you want me to include this point under the Drop-in directory for kubelet configuration files? Otherwise, it may not make much sense without that background knowledge

sohankunkerkar replied:
>In the example in this section, the feature gate consist of a single "AllAlpha": false. I wonder if it was configured by you explicitly or this is what it returns out of the box with nothing pre-configured?

I didn't configure the Kubelet config; it came out of the box.

SergeyKanzhelev replied:
> > what I am looking for when asking for /configz documentation is a clear statement on what that config represents:
> 
> > Does it merge-in all the command line arguments into it?
> > Does it merge-in all the drop-in configuration files?
> > Does it populate all the defaults?
> 
> Do you want me to include this point under the Drop-in directory for kubelet configuration files? Otherwise, it may not make much sense without that background knowledge

it does make sense when we talk about arguments and such. If it were separate page about /configz, than it will definitely make sense and we can simply refer to it. If we don't have it as of now - simply mentiuoning that this is a final merged config will make the most sense.

sohankunkerkar commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -84,45 +84,309 @@ If `--config` is provided and the values are not specified via the command line,
+
+#### Maps, including Nested Structures
+
+Individual fields in maps, regardless of their value types (boolean, string, etc.), can be selectively overridden. However, for `map[string][]string`, the entire list associated with a specific field gets overridden. Let's understand this better with an example, particularly on fields like `featureGates` and `staticPodURLHeader`:
```
ok, so the kubelet config by default sets the value for `AllAlpha` to false. It should be there in the final merged config. Let me fix this part.

SergeyKanzhelev commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -84,45 +84,310 @@ If `--config` is provided and the values are not specified via the command line,
 {{< /note >}}
 
-For instance, you may want a baseline kubelet configuration for all nodes, but you may want to customize the `address` field. This can be done as follows:
+The kubelet processes files in its config drop-in directory by sorting the **entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration files with a later sort order can override earlier files. This offers you flexibility in how you manage and combine kubelet configuration that comes from different sources. However, it's important to note that the beh ...
```
```suggestion
The kubelet processes files in its config drop-in directory by sorting the **entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration files with a later sort order can override earlier files. Files may contain partial configuration and may not be a valid config files by themselves, validation only performed on a final resulting configuration file. This offers you flexibility in how you manage and combine kubelet configuration that comes from different sources. However, it's important to note that the behavior varies based on the data type of the configuration fields. Here are some examples of how different data types behave during configuration merging:
```

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

fsmunoz commented:
Hello   @sohankunkerkar   :wave:  ! 

Please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before the deadline **Tuesday March 12th 2024 18:00 PST**. 

Thank you!

divya-mohan0209 commented:
@sohankunkerkar Since these are amendments for 1.30, it'd make sense to target the dev-1.30 branch. Please do the needful at the earliest possible.

sohankunkerkar commented:
> Hey @sohankunkerkar! I wanted to double-check if these changes in this pull request are specifically geared towards refining the current docs i.e. v1.29 docs. The [description](https://github.com/kubernetes/website/pull/44985#issue-2113442752) of the current PR seems a bit unclear in this regard. Are you planning to create a separate PR for the changes intended for `v1.30`? If that's the case, could you kindly update the [kubernetes/enhancements#3983](https://github.com/kubernetes/enhancements/issues/3983) issue with the PR number (maybe even raise a placeholder PR against `dev-1.30` now) to ensure that the release team is properly informed and tracking the correct pull request? Once this is sorted out, we can proceed with merging this one. Thanks!

Hey @dipesh-rawat, the plan is to update the main branch with the outdated documentation and then create a pull request targeting the 1.30-dev branch to promote this feature to beta.

sohankunkerkar commented:
Here's the PR targeting to the `dev-1.30` branch:  https://github.com/kubernetes/website/pull/45479

dipesh-rawat commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+
+To inspect the kubelet configuration, follow these steps:
+
+1. Start a proxy server using [`kubectl proxy`](/docs/reference/generated/kubectl/kubectl-commands#proxy) in your terminal.
```
```suggestion
1. Start a proxy server using [`kubectl proxy`](/docs/reference/kubectl/generated/kubectl_proxy/) in your terminal.
```

dipesh-rawat commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+1. Start a proxy server using [`kubectl proxy`](/docs/reference/generated/kubectl/kubectl-commands#proxy) in your terminal.
+
+```bash
+$ kubectl proxy
````
We [do not include the command prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt), so please remove it.

dipesh-rawat commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+Replace `<node-name>` with the actual name of your node:
+
+```bash
+$ curl -X GET http://127.0.0.1:8001/api/v1/nodes/<node-name>/proxy/configz | jq .
````
Same here, please remove the prompt.

celestehorgan commented:
Hi @sohankunkerkar!

Can you change this PR to be against `dev-1.30` rather than main? `main` gets archived (we open a new branch for it) for each version, so I'd like the change for this to come in as a part of 1.30 and not before. Small thing, but it'd be lovely, thanks :)

sohankunkerkar commented:
> Hi @sohankunkerkar!
> 
> Can you change this PR to be against `dev-1.30` rather than main? `main` gets archived (we open a new branch for it) for each version, so I'd like the change for this to come in as a part of 1.30 and not before. Small thing, but it'd be lovely, thanks :)

@celestehorgan There's a separate PR for tracking the 1.30-dev: #45479

sohankunkerkar commented:
@divya-mohan0209 could someone from the docs team review this PR?

drewhagen commented:
@kubernetes/sig-docs-en-reviews Hello! Could someone help give a review ahead of the March 26th Doc Freeze? I appreciate your help!

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+address: "192.168.0.8"
+```
+
+#### Lists/Slices
````
(nit)
```suggestion
#### Configuration merge behavior for lists and maps
``

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+The kubelet processes files in its config drop-in directory by sorting the **entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration files with a later sort order can override earlier files. These files may contain partial configurations and might not be valid config files by themselves; validation is only performed on the final resulting configura ...
+
+#### Singular/Embedded struct fields
+The merging process effectively handles the overriding of singular or embedded struct fields in kubelet configuration. For instance, you may want a baseline kubelet configuration for all nodes, but you may want to customize the `address` and `authorization` fields. This can be done as follows:
```
Readers likely have no idea what you mean by “embedded struct” (and we should not force them to know).

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
-For instance, you may want a baseline kubelet configuration for all nodes, but you may want to customize the `address` field. This can be done as follows:
+The kubelet processes files in its config drop-in directory by sorting the **entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration files with a later sort order can override earlier files. These files may contain partial configurations and might not be valid config files by themselves; validation is only performed on the final resulting configura ...
+
+#### Singular/Embedded struct fields
```
```suggestion
#### Configuration merge behavior for simple values
```

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+```
+
+#### Lists/Slices
+You can overidde the slices/lists values of the kubelet configuration, however, the entire list gets overridden during the merging process. For example, you can override the `clusterDNS` list as follows:
````
I wouldn't use “slice” here; YAML calls it a _mapping_.

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
Overall, there's a lot of detail here that belongs in a reference page.

Someone trying to follow the existing task page will find that task more confusing if we make these changes. Can we move the details elsewhere? For example, within https://kubernetes.io/docs/concepts/cluster-administration/?

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+
+```bash
+curl -X GET http://127.0.0.1:8001/api/v1/nodes/<node-name>/proxy/configz | jq .
+{
````
We aim to [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output).

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
````diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
+Replace `<node-name>` with the actual name of your node:
+
+```bash
+curl -X GET http://127.0.0.1:8001/api/v1/nodes/<node-name>/proxy/configz | jq .
````
If requiring `curl` and `jq`, make sure to list these as task prerequisites.

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
 
 This produces the same outcome as if you used the [single configuration file](#create-the-config-file) used in the earlier example.
 
+## Viewing the kubelet configuration
```
I would actually promote this to be its own task page (perhaps also its own commit, or own PR).

sftim commented on content/en/docs/tasks/administer-cluster/kubelet-config-file.md:
```diff
@@ -86,46 +86,310 @@ In the above example, this version is `kubelet.config.k8s.io/v1beta1`.
 
+## Viewing the kubelet configuration
+
+To inspect the kubelet configuration, follow these steps:
```
Do you need any access grants to be able to do this?

sftim reviewed (commented):
Thanks

I have some more feedback.

fsmunoz commented:
@sohankunkerkar hello, Release Lead Shadow for 1.30, acting as Docs shadow here; as @drewhagen mentioned above, the Docs Freeze is fast approaching (Tuesday March 26th 18:00 PDT).

From what I've seen, the technical review was started (@haircommander , @SergeyKanzhelev , if I'm not mistaken) and some of the items merged, but it's still not done - this needs to be done.

There are also some pending comments from SIG Docs - could you take a look? I think that with that (technical + Docs) the `lgtm` and `approve` process can follow, and hopefully before the freeze deadline. Thanks!

drewhagen commented:
@sohankunkerkar Just a heads up too, after the Doc Freeze deadline passes tomorrow, an [Exception needs to be filed](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) so that this feature can move forward releasing in 1.30. As a personal user of Kubernetes, I really appreciate your contribution to quality docs! 😄 PTAL at this tomorrow and let me know if there is anything we can do to help you over the finish line.

Thanks!!

haircommander commented:
I think the 1.30 PR covers this for docs freeze, right? TBH I am not sure we should move forward with this PR at this point

sftim commented:
It'd be nice to fixup the existing docs and the changes from https://github.com/kubernetes/website/pull/45665 so that they follow our style guides.

drewhagen commented:
/milestone 1.30

dipesh-rawat commented:
@sohankunkerkar What are our next steps for this PR? Is it in need of further work, ready for review, or can it be closed? Also, currently it needs a rebase against the _main_ branch.

sohankunkerkar commented:
All changes have been merged in the latest branch. I'm going to close this PR for now.
